src/s2/util/math/exactfloat/exactfloat.cc

```diff
@@ -96,15 +96,29 @@ inline static void BN_ext_set_uint64(BIGNUM* bn, uint64 v) {
 // Return the absolute value of a BIGNUM as a 64-bit unsigned integer.
 // Requires that BIGNUM fits into 64 bits.
 inline static uint64 BN_ext_get_uint64(const BIGNUM* bn) {
-  S2_DCHECK_LE(BN_num_bytes(bn), sizeof(uint64));
+  int bs = BN_num_bytes(bn);
+  S2_DCHECK_LE(bs, sizeof(uint64));
 #if BN_BITS2 == 64
   return BN_get_word(bn);
 #else
   static_assert(BN_BITS2 == 32, "at least 32 bit openssl build needed");
-  if (bn->top == 0) return 0;
-  if (bn->top == 1) return BN_get_word(bn);
-  S2_DCHECK_EQ(bn->top, 2);
+  if (bs == 0) return 0;
+  if (bs == 1) return BN_get_word(bn);
+  S2_DCHECK_EQ(bs, 2);
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
   return (static_cast<uint64>(bn->d[1]) << 32) + bn->d[0];
+#else  // OPENSSL_VERSION_NUMBER >= 0x10100000L
+  // In OpenSSL >= 1.1, BIGNUM is an opaque type, so d and top
+  // cannot be accessed.
+  BIGNUM *br = BN_new();
+  BN_rshift(br, bn, 32);
+  uint64_t ur = static_cast<uint64_t>(BN_get_word(br)) << 32;
```

Comment:
We're still using `uint64`, not `uint64_t`, and should switch everything together.